### PR TITLE
[2.19.x] Adds pipeline script for caching hero artifacts so they don't have to get rebuilt

### PR DESCRIPTION
**Pull request by Lambeaux**
#### What does this PR do?
Testing is important, but the current way we manually test makes it really difficult to batch up the tests as part of an individual developer process. The problem is that the developer has to rebuild the PR on their own before testing it. 
- If the developer uses their primary `.m2` then it will likely interfere with their ability to get work done, and so they've lost time equivalent to the build time.
- If the developer runs the build against a clean `.m2` to avoid polluting their own, they can still work, but it takes even longer due to dependencies re-downloading; I've often forgot I was even in the middle of a hero.

In my mind, this is not acceptable. If my team needs their stuff tested, I want that to be my priority, and I want to do every hero back-to-back to get it done, and then move on. We're already building PRs, why not just use those artifacts? 

This pipeline script is an early look at some automation that will cache the distribution zip from a PR build onto a Nexus with spare capacity and also handle the eviction of the cached artifacts. The idea is that performing a hero will simply require downloading the zip according to the project and PR number. 

Limitations:
- This cache won't be smart enough to build downstream PRs that depend on upstream PRs opened against dependencies.

#### Who is reviewing it? 
@LinkMJB 
@TonyMorrison 

#### Select relevant component teams: 
@codice/build 
@codice/continuous-integration 

#### Ask 2 committers to review/merge the PR and tag them here.
@rymach 

#### How should this be tested?
Either on the test Jenkins environment or with jenkinsfile-runner. 

#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Threat Dragon models
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

File: pipelines/Jenkinsfile
Comment (Lambeaux) [inline pipelines/Jenkinsfile:31]: Once this code gets moved to where it belongs, I suspect these flags won't be necessary anymore. They were just really useful for local testing.

File: pipelines/Jenkinsfile
Comment (Lambeaux) [inline pipelines/Jenkinsfile:33]: This is a valid parameter and will determine the maximum number of cached artifacts allowed in Nexus under the groupId `com.connexta.heroes`.

File: pipelines/Jenkinsfile
Comment (Lambeaux) [inline pipelines/Jenkinsfile:2]: I'm using some dependencies that will require approval in Jenkins administration. For testing, to ensure my upload and eviction logic was correct, I ran https://github.com/jenkinsci/jenkinsfile-runner with `--no-sandbox` enabled. Definitely not something you'd do in prod but I'm using it basically as a REPL to ensure I'm writing good code.

File: pipelines/Jenkinsfile
Comment (TonyMorrison) [reply pipelines/Jenkinsfile:33]: cool

File: pipelines/Jenkinsfile
Comment (rymach) [inline pipelines/Jenkinsfile:38]: are you planning to address these TODOs now or in a future PR?

File: pipelines/Jenkinsfile
Comment (rymach) [inline pipelines/Jenkinsfile:79]: might be able to use groovy's `'''` here

File: pipelines/Jenkinsfile
Comment (rymach) [inline pipelines/Jenkinsfile:63]: was anything like https://blog.sonatype.com/keeping-your-nexus-repositories-clean-just-got-easier considered?

File: pipelines/Jenkinsfile
Comment (rymach) [inline pipelines/Jenkinsfile:36]: have we considered using `mvn deploy` to deploy the ddf zip artifact to nexus?

File: pipelines/Jenkinsfile
Comment (rymach) [inline pipelines/Jenkinsfile:53]: how do we distinguish between multiple builds of the same pr?

Comment (rymach) [reply pipelines/Jenkinsfile:53]: also how do users discover find the zip url to download?

**Review by rymach — APPROVED** [review]

**Review by TonyMorrison — APPROVED** [review]

File: pipelines/Jenkinsfile
Comment (Lambeaux) [reply pipelines/Jenkinsfile:38]: This is a "bonus effort" for "if time permits". I probably should have made this a draft PR, in hindsight. 

I'm going to address these now. I don't want to create more work for us down the road.

File: pipelines/Jenkinsfile
Comment (Lambeaux) [reply pipelines/Jenkinsfile:79]: Nice tip - I'll take a look at the docs for that.

File: pipelines/Jenkinsfile
Comment (Lambeaux) [reply pipelines/Jenkinsfile:53]: >how do we distinguish between multiple builds of the same pr?

Good question. I was planning on taking a simple "overwrite" approach. The notion of PR build number wouldn't apply. 

>also how do users discover the zip url?

Will likely need to post that on the PR

File: pipelines/Jenkinsfile
Comment (Lambeaux) [reply pipelines/Jenkinsfile:63]: Huh. Nice find. I didn't know that existed. I'll look into it.

File: pipelines/Jenkinsfile
Comment (Lambeaux) [reply pipelines/Jenkinsfile:36]: What are the advantages of that? Only one I can think of is getting checksums uploaded for free. I just didn't think it was critical since we're keeping this behind the VPN. Also, on my local test environment, I'm testing out of the context of a build and didn't want to assume maven was accessible (even though it likely would be). 

I used `curl` to keep this as simple as possible. I think keeping it simple is the right route, but if there are other advantages I'm missing, I'm open to it.

File: pipelines/Jenkinsfile
Comment (Lambeaux) [reply pipelines/Jenkinsfile:63]: @rymach Yes, after talking with @LinkMJB we decided to go with a Nexus 3 cleanup policy. I'll clean up this PR next week.

File: pipelines/Jenkinsfile
Comment (rymach) [reply pipelines/Jenkinsfile:36]: deploy is less lines of code (`mvn deploy -pl distribution/ddf`) and handles credentials automatically (no hardcoding)

Comment (rymach) [reply pipelines/Jenkinsfile:36]: it also handles making the artifact names unique

**Comment by millerw8** [comment]
build now

**Comment by Lambeaux** [comment]
Just a heads up this PR is blocked on external resources.

**Pull request closed without merging**
